lgc/builder/BuilderImpl.cpp

```diff
@@ -658,7 +657,61 @@ static bool instructionsEqual(Instruction *lhs, Instruction *rhs) {
 
   return true;
 }
-#endif
+
+// =====================================================================================================================
+// Exttract the 32-bit value of the non-uniform index.
+// @param nonUniformIndex : the non-uniform index of the non-uniform operand of the image call
+// @return : the 32-bit value of the nonUniformIndex
+Value *get32BitNonUniformIndex(Value *nonUniformIndex) {
+  if (nonUniformIndex->getType()->isIntegerTy(64)) {
+    auto sExt = dyn_cast<SExtInst>(nonUniformIndex);
+    if (sExt && sExt->getOperand(0)->getType()->isIntegerTy(32))
+      return sExt->getOperand(0);
+    else {
+      Type *Int32Ty = Type::getInt32Ty(nonUniformIndex->getType()->getContext());
+      IRBuilder<> Builder(cast<Instruction>(nonUniformIndex)->getNextNode());
+      return Builder.CreateTrunc(nonUniformIndex, Int32Ty);
+    }
+  }
+  return nonUniformIndex;
+}
+
+// =====================================================================================================================
+// Checks if all the operands of the image call have a shared non-uniform index and returns the shared non-uniform
+// index.
+// @param nonUniformIndices : the non-uniform indices of the non-uniform operands of the image call
+// @param nonUniformIndex32BitVal : the map between the non-uniform index and its 32-bit value
+// @param traceNonUniformIndex : it is used to check if the non-uniform index can be scalarized
+// @param nonUniformInst : image call
+// @return : the 32-bit value of the nonUniformIndex
+Value *getSharedIndex(ArrayRef<Value *> nonUniformIndices, DenseMap<Value *, Value *> &nonUniformIndex32BitVal,
+                      TraceNonUniformIndex &traceNonUniformIndex, Instruction *nonUniformInst) {
```

Comment:
Unused `nonUniformInst`?
(However the function declares a local variable nonUniformInst**r**).

Reply:
Thank you for spotting it. Now, I renamed this function to calculateSharedIndex() and it is a method of TraceNonUniformIndex.